hyperlink/_url.py

```diff
@@ -481,13 +504,26 @@ def _percent_decode(text, _decode_map=_HEX_CHAR_MAP):
     res = [bits[0]]
     append = res.append
 
-    for item in bits[1:]:
-        try:
-            append(_decode_map[item[:2]])
-            append(item[2:])
-        except KeyError:
-            append(b'%')
-            append(item)
+    if not normalize_case:
+        for item in bits[1:]:
+            try:
+                append(_decode_map[item[:2]])
+                append(item[2:])
+            except KeyError:
+                append(b'%')
+                append(item)
+    else:
+        for item in bits[1:]:
+            try:
+                append(_decode_map[item[:2]])
+                append(item[2:])
+            except KeyError:
+                append(b'%')
+                if item[:2] in _HEX_CHAR_MAP:
+                    append(item[:2].upper())
+                    append(item[2:])
+                else:
+                    append(item)
```

Comment:
Ah, yes, because `_percent_decode` is called so frequently, I did a bit of optimization there. Checking the branch just once has a measurable effect on longer percent-encoded strings (and I've seen some pretty big ones). I can add a comment to this effect.